### ISSUE TITLE
Documentation: avoid pittfall of modifying a QueryBuilder instance (or call it with variable property values).

### ISSUE THREAD

**Issue by GerritDeMeulder** [objectbox/objectbox-java#223]
## Issue Basics
- ObjectBox version : 1.1.0
- Reproducibility: always

## Reproducing the bug
Run a project with 2 calls to a Querybuilder, app will not create a new modified object-instance , but will crash on second call.
Also important in loops, if you don't use setParameter() to modify the criteria, crash is similar.
### Description
If you have a `QueryBuilder `instance, and use it in a loop with variable property values or at a later point after `.build()`  try to redefine it to build a new query instance on the same entities group, but have overseen you used it already - and should use `setparameter()` method to modify it or create a new instance , you will get a JNI runtime NPE. 

### Code

```
private QueryBuilder<Term> termsQueryBuilder;
...
 termsQueryBuilder
        .equal(Term_.value, "")
        .equal(Term_.type, 0);
 Query<Term> termQuery = termsQueryBuilder.build();
 termQuery
       .setParameter(Term_.value, searchValue)
       .setParameter(Term_.type, selectedType); // This is the correct way !!
 term = termQuery.findFirst();
...
for (Term termToFind: termsMeetingCriteria){
...
termsQueryBuilder
      .greater(Term_.frequency, termToFind.frequency)
      .equal(Term_.type, termToFind.type);
List<Term> terms = termsQueryBuilder.build().find(); // Here it will crash (also without above part)!
...
}
```
### Logs & stackstraces
__runtime error__: tombstone in the likes of
```
Build fingerprint: 'Android/sdk_google_phone_x86_64/generic_x86_64:6.0/MASTER/4309849:userdebug/test-keys'
Revision: '0'
ABI: 'x86'
pid: 7257, tid: 7627, name: Thread-57  >>> com.company.myapp <<<
signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x0
    eax 00000000  ebx e089ab5c  ecx b997cad4  edx 00000000 
...
```
-> SIGSEGV error with addres 0 above is a Null Pointer Exception in Objectbox's native C-code, as far as I understand it.
## Misc
Did you find any workarounds to prevent the issue? 
Yes : use `setParameter()` instead of redefining the object instance.

**Comment by greenrobot** [comment]
QueryBuilder is intended to be used once at the moment. However, it should fail more gracefully than this... ;-)

**Comment by greenrobot** [comment]
I just checked calling build() twice on the same builder, and I'm getting this:

    java.lang.IllegalStateException: This QueryBuilder has already been closed. Please use a new instance.

This is expected behavior, and the question now is how did you trick the QueryBuilder into not throwing this exception?

**Comment by GerritDeMeulder** [comment]
@greenrobot : Good question, I know that by using trial and error, just before I solved the error,  I have seen that 
> java.lang.IllegalStateException: This QueryBuilder has already been closed. Please use a new instance.

also, that hinted me to the solution. 
I had to go to my Git commits, and I have no intermediate commit that contains the error. But I had another part where it's easy to create a similar error.

I believe it's really when you use it with parameters in a `for` loop, 
a normal second call will give you above `IllegalStateException`.
Because when I walk through it in debugging, it crashes at the point where the second chained condition method (i.e. `.equal(Term_.type, termToFind.type);`) is called in second call of the loop
So the isolated pattern to recreate the issue is : 
```
private QueryBuilder<Term> termsQueryBuilder;
...
for (Term termToFind: termsMeetingCriteria){
...
    termsQueryBuilder
      .greater(Term_.frequency, termToFind.frequency)
      .equal(Term_.type, termToFind.type);
     List<Term> terms = termsQueryBuilder.build().find(); 
...
}
```
NB:  I assume the following is not the  root cause : I am setting the link ` termsQueryBuilder = termsBox.query();` in the activity's `onCreate` method and executing the actual query on a background thread  (posted with EventBus) ;-) ), but since the other queries pose no problem, I assume this is not the  root cause.
